src/codegen/wasm.zig

```diff
@@ -305,56 +314,52 @@ pub const Context = struct {
         return WValue{ .local = self.local_index };
     }
 
-    fn genAdd(self: *Context, inst: *Inst.BinOp) InnerError!WValue {
+    fn genBinOp(self: *Context, inst: *Inst.BinOp, op: wasm.Op) InnerError!WValue {
         const lhs = self.resolveInst(inst.lhs);
         const rhs = self.resolveInst(inst.rhs);
 
         try self.emitWValue(lhs);
         try self.emitWValue(rhs);
 
-        const opcode: wasm.Opcode = switch (inst.base.ty.tag()) {
-            .u32, .i32 => .i32_add,
-            .u64, .i64 => .i64_add,
-            .f32 => .f32_add,
-            .f64 => .f64_add,
-            else => return self.fail(inst.base.src, "TODO - Implement wasm genAdd for type '{s}'", .{inst.base.ty.tag()}),
-        };
-
+        const opcode: wasm.Opcode = wasm.buildOpcode(.{
+            .op = op,
+            .valtype1 = try self.typeToValtype(inst.base.src, inst.base.ty),
+        });
         try self.code.append(wasm.opcode(opcode));
         return .none;
     }
 
     fn emitConstant(self: *Context, inst: *Inst.Constant) InnerError!void {
         const writer = self.code.writer();
-        switch (inst.base.ty.tag()) {
-            .u32 => {
-                try writer.writeByte(wasm.opcode(.i32_const));
-                try leb.writeILEB128(writer, inst.val.toUnsignedInt());
-            },
-            .i32, .bool => {
-                try writer.writeByte(wasm.opcode(.i32_const));
-                try leb.writeILEB128(writer, inst.val.toSignedInt());
+        switch (inst.base.ty.zigTypeTag()) {
+            .Int, .Bool, .Float => {
+                // write opcode
+                const opcode: wasm.Opcode = wasm.buildOpcode(.{
+                    .op = .@"const",
+                    .valtype1 = try self.typeToValtype(inst.base.src, inst.base.ty),
+                });
+                try writer.writeByte(wasm.opcode(opcode));
+                // write constant
+                switch (inst.base.ty.tag()) {
+                    .u32, .u64 => {
+                        try leb.writeILEB128(writer, inst.val.toUnsignedInt());
+                    },
+                    .i32, .i64, .bool => {
+                        try leb.writeILEB128(writer, inst.val.toSignedInt());
+                    },
+                    .f32 => {
+                        // TODO: enforce LE byte order
+                        try writer.writeAll(mem.asBytes(&inst.val.toFloat(f32)));
+                    },
+                    .f64 => {
+                        // TODO: enforce LE byte order
+                        try writer.writeAll(mem.asBytes(&inst.val.toFloat(f64)));
+                    },
+                    else => unreachable,
```

Comment:
There are more ways to represent Int, Bool, and Float than these tags. Instead of switching on tag() you should use the higher level API functions such as `intInfo`, `floatBits`.

Reply:
Done.

Note to self: can't combine the 3 cases (Int, Bool, Float) because intInfo doesn't work on Bool and Float, floatInfo doesn't work on Int and Bool.